### ISSUE TITLE
Patch files not included in mirror

### ISSUE THREAD

**Issue by louisvernon** [spack/spack#5344]
When creating a mirror spack does not seem to include required patch files.

On machine with internet access:
``` 
$ spack mirror create nauty
==> Adding package nauty@2.6r7 to mirror
==> Fetching http://pallini.di.uniroma1.it/nauty26r7.tar.gz
######################################################################## 100.0%
==> nauty@2.6r7 : checksum passed
==> nauty@2.6r7 : added
==> Successfully created mirror in spack-mirror-2017-09-12
  Archive stats:
    0    already present
    1    added
    0    failed to fetch.
```

On machine without internet access:
```
==> Installing nauty
==> Fetching file:///my/local/mirror/spack/mirror/nauty/nauty-2.6r7.tar.gz
######################################################################## 100.0%
==> Staging archive: /my/local/mirror/spack/var/spack/stage/nauty-2.6r7-attii264nx6pydjmpvwpnwzhlca23k5i/nauty26r7.tar.gz
==> Created stage in /my/local/mirror/spack/var/spack/stage/nauty-2.6r7-attii264nx6pydjmpvwpnwzhlca23k5i
==> Fetching file:///my/local/mirror/spack/mirror/nauty/nauty-fix-gt_numorbits.patch
curl: (37) Couldn't open file /my/local/mirror/spack/mirror/nauty/nauty-fix-gt_numorbits.patch
==> Fetching from file:///my/local/mirror/spack/mirror/nauty/nauty-fix-gt_numorbits.patch failed.
==> Fetching https://src.fedoraproject.org/rpms/nauty/raw/0f07d01caf84e9d30cb06b11af4860dd3837636a/f/nauty-fix-gt_numorbits.patch

curl: (22) The requested URL returned error: 403
==> Fetching from https://src.fedoraproject.org/rpms/nauty/raw/0f07d01caf84e9d30cb06b11af4860dd3837636a/f/nauty-fix-gt_numorbits.patch failed.
==> Patch https://src.fedoraproject.org/rpms/nauty/raw/0f07d01caf84e9d30cb06b11af4860dd3837636a/f/nauty-fix-gt_numorbits.patch failed.
==> Error: FetchError: All fetchers failed for spack-stage-e2ZX5W
FetchError: FetchError: All fetchers failed for spack-stage-e2ZX5W

/turquoise/usr/projects/ic/spack/lib/spack/spack/package.py:1029, in do_patch:
     50                            # Touch bad file if anything goes wrong.
     51                            tty.msg('Patch %s failed.' % patch.path_or_url)
     52                            touch(bad_file)
  >> 53                            raise
     54    
     55            if has_patch_fun:
     56                try:
```

**Comment by scheibelp** [comment]
I'll be able to look at this on 9/18

**Comment by junghans** [comment]
@scheibelp: NWChem has the same problem.

I think one needs to do something like:
```python
   resource(
       name='nauty-fix-gt_numorbits.patch',
       url='https://src.fedoraproject.org/rpms/nauty/raw/0f07d01caf84e9d30cb06b11af4860dd3837636a/f/nauty-fix-gt_numorbits.patch',  # noqa: E50
       md5='a6e1ef4897aabd67c104fd1d78bcc334',
       destination='.',
       placement='.',
       expand=False,
       when='@2.6r7')
    patch('nauty-fix-gt_numorbits.patch', when='@2.6r7')
```

**Comment by scheibelp** [comment]
It appears at the moment we don't support this other than via the method suggested by @junghans. I'll note that method is a bit confusing because:

1. resources are typically added to the source files but patches don't need to be
2. this actually avoids the effect described in (1) by setting `destination` and `placement` to `.` (which has an unexpected effect that the patch file is not placed in the source directory)

So although a rough workaround exists now, I ought to address the lack of support for this in the patching functionality. I'd like to wait until #3553 gets merged (@tgamblin, @becker33) to work on that.

**Comment by IanLee1521** [comment]
Just ran into this myself with perl not pulling in the patch file (see: https://github.com/spack/spack/blob/76ab54caec803e55034b531401e277dfd51ffe80/var/spack/repos/builtin/packages/perl/package.py#L62) -- Discussed with @becker33

**Comment by scheibelp** [comment]
This should be addressed by #8993 (not merged yet)
